l1/l1.go

```diff
@@ -0,0 +1,141 @@
+package l1
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/l1/contract"
+	"github.com/NethermindEth/juno/service"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+//go:generate mockgen -destination=./mocks/mock_subscriber.go -package=mocks github.com/NethermindEth/juno/l1 Subscriber
+type Subscriber interface {
+	WatchHeader(ctx context.Context, sink chan<- *types.Header) (event.Subscription, error)
+	WatchLogStateUpdate(ctx context.Context, sink chan<- *contract.StarknetLogStateUpdate) (event.Subscription, error)
+}
+
+type Client struct {
+	l1                Subscriber
+	l2Chain           *blockchain.Blockchain
+	log               utils.SimpleLogger
+	confirmationQueue *queue
+}
+
+var _ service.Service = (*Client)(nil)
+
+func NewClient(l1 Subscriber, chain *blockchain.Blockchain, confirmationPeriod uint64, log utils.SimpleLogger) *Client {
+	return &Client{
+		l1:                l1,
+		l2Chain:           chain,
+		log:               log,
+		confirmationQueue: newQueue(confirmationPeriod),
+	}
+}
+
+func (c *Client) subscribeToHeaders(ctx context.Context, headerChan chan *types.Header) (event.Subscription, error) {
+	sub, err := c.l1.WatchHeader(ctx, headerChan)
+	if err != nil {
+		return nil, fmt.Errorf("subscribe to L1 headers: %w", err)
+	}
+	return sub, nil
+}
+
+func (c *Client) subscribeToUpdates(ctx context.Context,
+	logStateUpdateChan chan *contract.StarknetLogStateUpdate,
+) (event.Subscription, error) {
+	sub, err := c.l1.WatchLogStateUpdate(ctx, logStateUpdateChan)
+	if err != nil {
+		return nil, fmt.Errorf("subscribe to L1 state updates: %w", err)
+	}
+	return sub, nil
+}
+
+func (c *Client) Run(ctx context.Context) error {
+	buffer := 128
+
+	logStateUpdateChan := make(chan *contract.StarknetLogStateUpdate, buffer)
+	subUpdates, err := c.subscribeToUpdates(ctx, logStateUpdateChan)
+	if err != nil {
+		return err
+	}
+	defer subUpdates.Unsubscribe()
+
+	headerChan := make(chan *types.Header, buffer)
+	subHeaders, err := c.subscribeToHeaders(ctx, headerChan)
+	if err != nil {
+		return err
+	}
+	defer subHeaders.Unsubscribe()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-subHeaders.Err():
+			c.log.Warnw("L1 header subscription failed, resubscribing", "error", err)
+			subHeaders.Unsubscribe()
```

Comment:
Are you sure that `Unsubscribe()` won't close the `headerChan`?

Reply:
The documentation says `Unsubscribe()` "cancels sending of events, closing the error channel". It should not close the channels it did not create.

Reading the implementation, I can confirm `Unsubscribe()` doesn't close the `headerChan`.